### ISSUE TITLE
0.6.5.4 deadlock during snapshot umount

### ISSUE THREAD

**Issue by alexanderhaensch** [zfsonlinux/zfs#4379]
the automounted snapshots in .zfs/snapshot do not unmount. A deadlock is created instead

kernel:
 4.1.15-gentoo-r1

spl/zfs:
0.6.5.4

dmesg:
```
INFO: rcu_sched self-detected stall on CPU
 2: (3607981 ticks this GP) idle=d95/140000000000001/0 softirq=166024/166137 fqs=3416341 
  (t=3607982 jiffies g=168810 c=168809 q=22167456)
Task dump for CPU 2:
umount          R  running task        0 24873  24872 0x00000008
 ffff88201d372a80 000000005d4b0605 ffff88103fc43d98 ffffffff810759ce
 0000000000000002 ffffffff81a391c0 ffff88103fc43db8 ffffffff81078e68
 ffff88103fc43db8 0000000000000003 ffff88103fc43de8 ffffffff810988b1
Call Trace:
 <IRQ>  [<ffffffff810759ce>] sched_show_task+0xae/0x120
 [<ffffffff81078e68>] dump_cpu_task+0x38/0x40
 [<ffffffff810988b1>] rcu_dump_cpu_stacks+0x91/0xd0
 [<ffffffff8109bb03>] rcu_check_callbacks+0x433/0x710
 [<ffffffff810d8737>] ? acct_account_cputime+0x17/0x20
 [<ffffffff810ad230>] ? tick_sched_handle.isra.18+0x40/0x40
 [<ffffffff8109df94>] update_process_times+0x34/0x60
 [<ffffffff810ad221>] tick_sched_handle.isra.18+0x31/0x40
 [<ffffffff810ad26c>] tick_sched_timer+0x3c/0x80
 [<ffffffff8109ea21>] __run_hrtimer.isra.34+0x41/0xd0
 [<ffffffff8109ecc5>] hrtimer_interrupt+0xd5/0x200
 [<ffffffff81037187>] local_apic_timer_interrupt+0x37/0x60
 [<ffffffff8103783c>] smp_apic_timer_interrupt+0x3c/0x50
 [<ffffffff8168153b>] apic_timer_interrupt+0x6b/0x70
 <EOI>  [<ffffffff81680393>] ? _raw_spin_unlock_irqrestore+0x13/0x20
 [<ffffffffa06bd091>] taskq_wait_outstanding+0x51/0x100 [spl]
 [<ffffffffa08387f0>] zfs_sb_teardown+0x60/0x4c0 [zfs]
 [<ffffffffa083a49f>] zfs_umount+0x2f/0x140 [zfs]
 [<ffffffffa086522d>] zpl_put_super+0x2d/0xc0 [zfs]
 [<ffffffff81141311>] generic_shutdown_super+0x71/0x100
 [<ffffffff81141631>] kill_anon_super+0x11/0x20
 [<ffffffffa08654e9>] zpl_kill_sb+0x19/0x20 [zfs]
 [<ffffffff81141a0c>] deactivate_locked_super+0x3c/0x70
 [<ffffffff81141e6c>] deactivate_super+0x6c/0x80
 [<ffffffff8115e7ce>] cleanup_mnt+0x3e/0x90
 [<ffffffff8115e86d>] __cleanup_mnt+0xd/0x10
 [<ffffffff8106b214>] task_work_run+0xd4/0xf0
 [<ffffffff81003b98>] do_notify_resume+0x68/0x70
 [<ffffffff81680942>] int_signal+0x12/0x17
```

It seems that two umounts locking each other:
```

root     24872  0.0  0.0  18256  2712 ?        S    Feb28   0:00 /bin/sh -c exec 0</dev/null      1>/dev/null      2>/dev/null; umount -t zfs -n '/storage/wingroups/.zfs/snapshot/zfs-auto-snap_05-2016-02-04-0210'
root     24873  100  0.0  11012  2108 ?        R    Feb28 673:22 umount -t zfs -n /storage/wingroups/.zfs/snapshot/zfs-auto-snap_05-2016-02-04-0210
root     26331  0.0  0.0  18256  2752 ?        S    Feb28   0:00 /bin/sh -c exec 0</dev/null      1>/dev/null      2>/dev/null; umount -t zfs -n -f '/storage/wingroups/.zfs/snapshot/zfs-auto-snap_04-2016-02-27-2000'
root     26332  0.0  0.0  11012  2124 ?        D    Feb28   0:00 umount -t zfs -n -f /storage/wingroups/.zfs/snapshot/zfs-auto-snap_04-2016-02-27-2000
```

**Comment by alexanderhaensch** [comment]
This does not happen with 0.6.5.3.
could this be a sideeffect of: https://github.com/zfsonlinux/zfs/commit/3445a340d54f4209eb13cf9a59b01ef571006fc4

**Comment by alexanderhaensch** [comment]
The situation can get worse if i kill an umount process, for example 24872. I get another dead process.

zfs destroy on a dataset does deadlock too :(

**Comment by lazy404** [comment]
this seems to be the same issue as https://github.com/zfsonlinux/zfs/issues/4359
what are the exact timestamps of thiese processes ?

did disabling xattr help ?

in my case xattr is enabled

**Comment by alexanderhaensch** [comment]
@lazy404 ideed, thats the same issue.
I restartet the system without https://github.com/zfsonlinux/zfs/commit/3445a340d54f4209eb13cf9a59b01ef571006fc4 . But in the moment i have problems to recreate the issue, because the external app that produces the issue is not working right now.

**Comment by alexanderhaensch** [comment]
I cannot reproduce the issue without https://github.com/zfsonlinux/zfs/commit/3445a340d54f4209eb13cf9a59b01ef571006fc4 . I was changing something on the producing app. So maybe the opening logic was changed.
I think the issue is produced if concurrent access to multiple snapshot dirs is happening.

**Comment by ryao** [comment]
This is not caused by the stable API extensions.

**Comment by behlendorf** [comment]
Yes, this is currently a regression in the master source.  The good news is that it's fairly easy to reproduce with the ZFS Test Suite and causes quite a few test failures so it's on the short list of things to address.

**Comment by alexanderhaensch** [comment]
This is solved somewhere between 0.6.5.4 and 0.6.5.6 for me. Please open again if it still affects you.
